### PR TITLE
[KED-1555] Expose VERSION_FORMAT constant in Kedro IO library

### PR DESCRIPTION
**Pull request by deepyaman**
## Description
Replaces #299 (broken fork).

I’m currently trying to parse versions, and wanted to see if I could avoid hardcoding the version format myself.

This is my code:
```python3
from datetime import datetime
last_load_version = datetime.strptime(
    catalog.datasets.data_set_name.get_last_load_version(),  # Replace with `resolve_load_version` in future version of Kedro
    "%Y-%m-%dT%H.%M.%S.%fZ",
)
last_load_version
```

## Development notes
Manual test:
```python3
>>> from datetime import datetime
>>> 
>>> from kedro.io.core import VERSION_FORMAT
>>> 
>>> datetime.strptime(generate_timestamp(), VERSION_FORMAT)
datetime.datetime(2020, 3, 21, 21, 33, 59, 405000)
>>> 
```

~~I'll add a unit test verifying converted-back timestamp is correct (no microseconds) later; I'm currently having issues with local tests and CircleCI... =/ (all local test failures seemed completely unrelated)~~ Test added.

## Checklist

- [x] Read the [contributing](../CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](../RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: kedro/io/core.py [hunk 1]
```diff
@@ -334,12 +335,8 @@ def generate_timestamp() -> str:
         String representation of the current timestamp.
 
     """
-    current_ts = datetime.now(tz=timezone.utc)
-    fmt = (
-        "{d.year:04d}-{d.month:02d}-{d.day:02d}T{d.hour:02d}"
-        ".{d.minute:02d}.{d.second:02d}.{ms:03d}Z"
-    )
-    return fmt.format(d=current_ts, ms=current_ts.microsecond // 1000)
+    current_ts = datetime.now(tz=timezone.utc).strftime(VERSION_FORMAT)
```
[the remark this answers is not in the document]
Comment (lorenabalan) [reply kedro/io/core.py:338]: I know the timezone was already here in previous implementation, but are we actually making use of it anywhere?  Don't we lose it when we convert to string currently?

[the remark this answers is not in the document]
Comment (limdauto) [reply kedro/io/core.py:338]: The string value will be the time value in UTC though, so we don't actually lose it.

[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/io/core.py:338]: `Z` at the end of `VERSION_FORMAT` indicates that it's actually a UTC timestamp as per [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Coordinated_Universal_Time_(UTC))

Python is pretty bad at converting timezones into ISO format:
```python
datetime.now(tz=timezone.utc).isoformat(timespec="milliseconds")
# => 2020-04-16T16:11:29.894+00:00
```
Therefore I did it manually.

[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/io/core.py:338]: > or make `get_timestamp()` return `isoformat(timespec="milliseconds")` and then easily convert them back with `datetime.fromisoformat()`

`datetime.fromisoformat()` is [Python 3.7+](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) only.

**Comment by DmitriiDeriabinQB** [comment]
> or make `get_timestamp()` return `isoformat(timespec="milliseconds")` and then easily convert them back with `datetime.fromisoformat()`

@lorenabalan `datetime.fromisoformat()` is [Python 3.7+](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) only.

**Comment by lorenabalan** [comment]
> @lorenabalan `datetime.fromisoformat()` is [Python 3.7+](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) only.

Yeah I know it was just an example. There are other ways to deal with older versions though not as clean.

**Comment by DmitriiDeriabinQB** [comment]
Let me comment on a thought process that defined my choice for the current version format:

1. Versions need to be sequential (so GUIDs or other randoms don't work)
2. Version generation needs to be stateless, so that we don't share any global version state
3. It ideally needs to be human readable, hence timestamp with separators
4. It needs to be granular to reduce the chances of duplicate collisions, hence the addition of milliseconds (in case 2 runs start in parallel on the same machine within the same second)
5. It needs to be a filesystem-friendly, hence `:` replacement (not allowed in posix file/directory names) with `.` (fine in both posix and windows)
6. Version generation must be invariant to the environment (e.g., current timezone), hence UTC timestamp

File: kedro/io/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/io/core.py:338]: As discussed offline, we always generate a UTC timestamp regardless of the current timezone of the user machine, therefore this literal `Z` character indicates that it's a UTC time.
Absolutely right that `datetime.utcnow()` is legitimate here too. So if one calls `datetime.utcnow().isoformat(timespec="milliseconds").replace(":", ".") + "Z"` they'll get the same result as what `generate_timestamp()` currently returns.

[the remark this answers is not in the document]
Comment (DmitriiDeriabinQB) [reply kedro/io/core.py:338]: Oh, I now recalled the difference! `datetime.now(tz=timezone.utc)` produces timezone-aware datetime object, while `datetime.utcnow()` produces a naive one (which is super weird):
```python
x = datetime.utcnow()
x.tzinfo
# None
y = datetime.now(tz=timezone.utc)
y.tzinfo
# datetime.timezone.utc
```
It doesn't change anything in this particular use case, but naive timestamps are quite dangerous, imo.

**Comment by deepyaman** [comment]
Oops, I missed the whole discussion! 😂 Thanks for taking a look, all!

**Pull request merged**